crates/amaru-kernel/src/memoized.rs

```diff
@@ -0,0 +1,265 @@
+use pallas_addresses::Address;
+use pallas_codec::{minicbor as cbor, minicbor::data::IanaTag, utils::Bytes};
+use pallas_crypto::hash::{Hash, Hasher};
+use pallas_primitives::{
+    alonzo,
+    conway::{
+        KeepRaw, MintedDatumOption, MintedTransactionOutput, NativeScript, PseudoScript, Value,
+    },
+    DatumHash, NonEmptyKeyValuePairs, PlutusData,
+};
+
+// --------------------------------------------------- MemoizedTransactionOutput
+
+#[derive(Debug, Clone)]
+pub struct MemoizedTransactionOutput {
+    is_legacy: bool,
+    pub address: Address,
+    pub value: Value,
+    pub datum: MemoizedDatum,
+    pub script: Option<MemoizedScript>,
+}
+
+impl<'b, C> cbor::Decode<'b, C> for MemoizedTransactionOutput {
+    // TODO: This implementation still copies the bytes when we convert from the minted version to
+    // the owned version. To avoid that, we should actually rewrite the decoders for all the
+    // memoized structures, so that they can retain their original bytes.
+    fn decode(d: &mut cbor::Decoder<'b>, ctx: &mut C) -> Result<Self, cbor::decode::Error> {
+        let minted: MintedTransactionOutput<'_> = d.decode_with(ctx)?;
+        Self::try_from(minted).map_err(cbor::decode::Error::message)
+    }
+}
+
+impl<C> cbor::Encode<C> for MemoizedTransactionOutput {
+    fn encode<W: cbor::encode::Write>(
+        &self,
+        e: &mut cbor::Encoder<W>,
+        ctx: &mut C,
+    ) -> Result<(), cbor::encode::Error<W::Error>> {
+        if self.is_legacy {
+            e.begin_array()?;
+            e.bytes(&self.address.to_vec())?;
+            e.encode_with(&self.value, ctx)?;
+            match self.datum {
+                MemoizedDatum::None => (),
+                MemoizedDatum::Hash(hash) => {
+                    e.bytes(&hash[..])?;
+                }
+                MemoizedDatum::Inline(..) => unreachable!("legacy output with inline datum ?!"),
+            }
+            e.end()?;
+        } else {
+            e.begin_map()?;
+
+            e.u8(0)?;
+            e.bytes(&self.address.to_vec())?;
+
+            e.u8(1)?;
+            e.encode_with(&self.value, ctx)?;
+
+            match &self.datum {
```

Comment:
Perhaps it makes more sense to have the datum encoding belong in it's own implementation of `minicbor::Encode`?